src/sentry/runner/commands/upgrade.py

```diff
@@ -1,29 +1,134 @@
+    ("auth", "0006_require_contenttypes_0002"),
+    ("sites", "0001_initial"),
+    ("admin", "0001_initial"),
+    ("sessions", "0001_initial"),
+)
+
+
+def _has_applied_django_migration(connection, app_name, migration):
+    cursor = connection.cursor()
+    try:
+        cursor.execute(
+            "SELECT 1 FROM django_migrations WHERE app = %s AND name = %s LIMIT 1",
+            [app_name, migration],
+        )
+    except ProgrammingError as exc:
+        if 'relation "django_migrations" does not exist' in six.text_type(exc):
+            return False
+        raise
+    else:
+        return bool(cursor.fetchall())
+    finally:
+        cursor.close()
+
+
+def _has_applied_south_migration(connection, app_name, migration):
+    if migration is None:
+        return True
+    cursor = connection.cursor()
+    try:
+        cursor.execute(
+            "SELECT 1 FROM south_migrationhistory WHERE app_name = %s AND migration = %s LIMIT 1",
+            [app_name, migration],
+        )
+    finally:
+        cursor.close()
+        return bool(cursor.fetchall())
+
+
+def _fake_django_migration(connection, app_name, migration, verbosity=0):
+    if _has_applied_django_migration(connection, app_name, migration):
+        return False
+
+    if verbosity:
+        click.echo("Faking migration for {}.{}".format(app_name, migration))
+
+    cursor = connection.cursor()
+    try:
+        cursor.execute(
+            "INSERT INTO django_migrations (app, name, applied) VALUES (%s, %s, NOW())",
+            [app_name, migration],
+        )
+    finally:
+        cursor.close()
+
+
+def _has_south_history(connection):
+    cursor = connection.cursor()
+    try:
+        cursor.execute("SELECT 1 FROM south_migrationhistory LIMIT 1")
+    except ProgrammingError as exc:
+        if 'relation "south_migrationhistory" does not exist' in six.text_type(exc):
+            return False
+        raise
+    return True
+
+
+def _migrate_from_south(verbosity):
+    from django.db.migrations.recorder import MigrationRecorder
+
+    connection = connections["default"]
+
+    if not _has_south_history(connection):
+        return
+
+    # ensure the django_migrations relation exists
+    recorder = MigrationRecorder(connection)
+    recorder.ensure_schema()
+
+    for django_app_name, django_migration in DJANGO_MIGRATIONS:
+        _fake_django_migration(connection, django_app_name, django_migration, verbosity)
+
+    for (
+        south_app_name,
+        south_migration,
+        django_app_name,
+        django_migration,
+        south_migration_required,
+        south_migration_required_error,
+    ) in settings.SOUTH_MIGRATION_CONVERSIONS:
+        if _has_applied_south_migration(connection, south_app_name, south_migration):
+            _fake_django_migration(connection, django_app_name, django_migration)
+        elif south_migration_required:
+            raise Exception(south_migration_required_error)
```

Comment:
Shall we raise a more specific exception here like a DB exception: https://docs.djangoproject.com/en/1.8/ref/exceptions/#django.db.IntegrityError

Reply:
Can do, I mostly just want to display an error message to the user.

Reply:
I'd be careful on the kind of error - integrity error for example generally implies to anything looking for it that they should roll back.